### PR TITLE
Introduction of new functionality within the script "evaluate.py" to build and output the compiled model and its meta-information

### PR DESCRIPTION
**Pull request by dsbarinov1**
Motivation: A script "evaluate.py" is "used for now for testing of Qualcomm Adreno work" and it can handle inputs and outputs of a certain number of models. Part of the task "TO-19" in jira is to compile part of this range of models, so it would be convenient to reuse the avaliable functionality (via extending the existing one) from the "evaluate.py" script to solve the problem.
Changes:

- A new method "export_lib_and_meta" of class "ModelImporter" was added, which is used to compile models (without tuning/evaluating) by obtaining information (mod, params, shape_dict, target...) from "import_..(model_name).." methods, build process (for now only via Relay and Graph Executor) and to export compiled library (.so) and meta info (.txt) about compiled models;
- A new argument "--build_and_export" or "-BE" was created to explicitly indicate that we intend to compile and export the compiled library and meta information;
- A new argument "--output" or "-o" was created to specify the output path of the compiled library and meta information.

### CODE REVIEW

**Comment by dsbarinov1** [comment]
@KJlaccHoeUM9l

File: scripts/evaluate.py [hunk 1]
```diff
@@ -1240,16 +1309,19 @@ def get_args():
...
 def main():
-    if "opencl" in args.target:
-        executor = Executor(use_tracker="android")
-    else:
-        executor = Executor()
-    executor.schedule(args.model, target=args.target, dtype=args.type)
-    if args.tune:
-        executor.tune_pending_benchmarks()
+    if args.build_and_export:
+        ModelImporter().export_lib_and_meta(model_name=args.model, target="llvm", dtype="float32", output_dir=args.output)
```
Comment (KJlaccHoeUM9l) [inline scripts/evaluate.py:1313]: Use `args.target` and `args.dtype` instead of hardcoded values

File: scripts/evaluate.py [hunk 2]
```diff
@@ -1221,7 +1290,7 @@ def get_args():
         "log_filename": args.log,
         "early_stopping": None,
         "measure_option": autotvm.measure_option(
-            builder=autotvm.LocalBuilder(build_func=ndk.create_shared, timeout=15, n_parallel=2),
+            builder=autotvm.LocalBuilder(build_func=ndk.create_shared, timeout=15, n_parallel=1),
```
Comment (KJlaccHoeUM9l) [inline scripts/evaluate.py:1293]: For this task, tuning is not needed (yet), the code corresponding to tuning should not change

File: scripts/evaluate.py [hunk 3]
```diff
@@ -1149,6 +1149,69 @@ def validator(inputs):
...
+        import_method = name_to_model[model_name]
+        (mod, params, shape_dict, dtype, target, validator) = import_method(target, dtype)
+
+        with relay.build_config(opt_level=3):
+            graph, lib, params = relay.build(
+                mod, target_host=target, target=target, params=params
+            )
+
+        if args.output != './':
+            output_dir = output_dir + "/"
```
Comment (KJlaccHoeUM9l) [inline scripts/evaluate.py:1205]: This is not a universal solution. For example, on Windows, the path separator looks like this: `\`.
Delete this `if-else` block and use the `os.path.join` below:
```Python
lib.export_library(os.path.join(output_dir, model_name + "-default.so"))
```

File: scripts/evaluate.py [hunk 4]
```diff
@@ -1149,6 +1149,69 @@ def validator(inputs):
...
+        "mobilenetv1": self.import_mobilenetv1,
+        "mobilenetv3_ssdlite": self.import_mobilenetv3_ssdlite,
+        "resnet50": self.import_resnet50,
+        "resnet50_v2": self.import_resnet50_v2,
+        "vgg16": self.import_vgg16,
+        "yolov3_mxnet": self.import_yolov3_mxnet
+        }
+
+        import_method = name_to_model[model_name]
+        (mod, params, shape_dict, dtype, target, validator) = import_method(target, dtype)
```
Comment (KJlaccHoeUM9l) [inline scripts/evaluate.py:1197]: `validator` is not used, use `_` instead

File: scripts/evaluate.py [hunk 5]
```diff
@@ -1149,6 +1149,69 @@ def validator(inputs):
...
+            graph, lib, params = relay.build(
+                mod, target_host=target, target=target, params=params
+            )
+
+        if args.output != './':
+            output_dir = output_dir + "/"
+        lib.export_library(output_dir + model_name + "-default.so")
+        print("lib exported successfully!")
+
+        import json
```
Comment (KJlaccHoeUM9l) [inline scripts/evaluate.py:1209]: move `import` to the top of the file

File: scripts/evaluate.py [hunk 6]
```diff
@@ -1149,6 +1149,69 @@ def validator(inputs):
...
+
+        with open(output_dir + str(model_name) + '_data.txt', 'w+') as f:
```
Comment (KJlaccHoeUM9l) [inline scripts/evaluate.py:1211]: use `os.path.join`

File: scripts/evaluate.py [hunk 7]
```diff
@@ -1149,6 +1149,69 @@ def validator(inputs):
...
+            f.write(model_name + " meta info\nShape_dict = " + json.dumps(shape_dict) + "\nDtype = " + str(dtype) +"\nTarget = " + str(target))
+            f.close()
```
Comment (KJlaccHoeUM9l) [inline scripts/evaluate.py:1213]: `f.close()` will be called automatically after exiting scope of `context manager`, delete this line

File: scripts/evaluate.py [hunk 8]
```diff
@@ -1149,6 +1149,69 @@ def validator(inputs):
...
+
+        if args.output != './':
+            output_dir = output_dir + "/"
+        lib.export_library(output_dir + model_name + "-default.so")
+        print("lib exported successfully!")
+
+        import json
+
+        with open(output_dir + str(model_name) + '_data.txt', 'w+') as f:
+            f.write(model_name + " meta info\nShape_dict = " + json.dumps(shape_dict) + "\nDtype = " + str(dtype) +"\nTarget = " + str(target))
```
Comment (KJlaccHoeUM9l) [inline scripts/evaluate.py:1212]: use `json.dums` for all data

File: scripts/evaluate.py (hunk 3 above)
Comment (vvchernov) [reply scripts/evaluate.py:1205]: One more thing args.output is "./" by default which is Linux format. Nevertheless please use os.path or pathlib libraries for path processing. In the latter case you do not need to follow the path concatination and condition can be skipped:
`lib.export_library(os.path.join(args.output, model_name + "-default.so"))`

**Comment by dsbarinov1** [comment]
@KJlaccHoeUM9l

File: scripts/evaluate.py [hunk 9]
```diff
@@ -1149,6 +1150,75 @@ def validator(inputs):
...
+        (mod, params, shape_dict, dtype, target, _) = import_method(target, dtype)
+
+        with tvm.transform.PassContext(opt_level=3):
+            lib = relay.build(mod, target=target, params=params)
+
+        lib.export_library(os.path.join(output_dir, model_name + "-default.so"))
+        print("lib exported successfully!")
+
+        with open(os.path.join(output_dir, model_name + '_data.txt'), 'w+') as f:
+            f.write(json.dumps(model_name) + " meta info\nShape_dict = " + json.dumps(shape_dict) + "\nDtype = " + json.dumps(dtype) +"\nTarget = " + json.dumps(target))
```
Comment (KJlaccHoeUM9l) [inline scripts/evaluate.py:1220]: Use just one time:
```Python
json.dumps(
    {
         "Shape_dict": shape_dict,
         "Dtype": dtype,
         "Target": target,
     }
)
```

File: scripts/evaluate.py [hunk 10]
```diff
@@ -1213,43 +1289,51 @@ def get_args():
...
+                    #min_repeat_ms=150,
+                    #cooldown_interval=150
+                ),
             ),
-        ),
-    }
+        }
+    else:
+        args.tuning_options = {             
+        }
```
Comment (KJlaccHoeUM9l) [inline scripts/evaluate.py:1316]: why did you wrap all parameters?
the problem was only with `tuning_options`, you only need to wrap it.

File: scripts/evaluate.py [hunk 11]
```diff
@@ -1754,4 +1838,4 @@ def tune_tasks(
         # os.remove(tmp_log_file)
 
 if __name__ == "__main__":
-    main()
+    main()
```
Comment (KJlaccHoeUM9l) [inline scripts/evaluate.py:1841]: there shouldn't be any changes

**Pull request closed without merging**